### PR TITLE
Fix Segmentation Fault Crash

### LINKED ISSUE #39: Segmentation fault crash
Watcher has been crashing on macOS (10.14/10.15) with a segmentation fault. Should be easily reproduced by watching a large javascript repo and running `rm -rf node_modules && yarn` a number of times. I built a debug version of node 12.18.1 and a debug version of the watcher and was able to catch these stack traces:

[1]
```
Process 48323 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
    frame #0: 0x000000010600e154 watcher.node`bool std::__1::__tree_is_left_child<std::__1::__tree_node_base<void*>*>(__x=0x00000001039a6f60) at __tree:75:16
   72  	bool
   73  	__tree_is_left_child(_NodePtr __x) _NOEXCEPT
   74  	{
-> 75  	    return __x == __x->__parent_->__left_;
   76  	}
   77
   78  	// Determines if the subtree rooted at __x is a proper red black subtree.  If
Target 0: (node) stopped.
Process 48323 launched: '/Users/mteegarden/dev/repos/node/node' (x86_64)
(llnode) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x0)
  * frame #0: 0x000000010600e154 watcher.node`bool std::__1::__tree_is_left_child<std::__1::__tree_node_base<void*>*>(__x=0x00000001039a6f60) at __tree:75:16
    frame #1: 0x000000010600e0bf watcher.node`std::__1::__tree_end_node<std::__1::__tree_node_base<void*>*>* std::__1::__tree_next_iter<std::__1::__tree_end_node<std::__1::__tree_node_base<void*>*>*, std::__1::__tree_node_base<void*>*>(__x=0x00000001039a6f60) at __tree:178:13
    frame #2: 0x000000010600e06f watcher.node`std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event>, void*>*, long>::operator++(this=0x00007ffeefbf6bd8) at __tree:839:11
    frame #3: 0x000000010600e039 watcher.node`std::__1::__map_iterator<std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event>, void*>*, long> >::operator++(this=0x00007ffeefbf6bd8) at map:803:35
    frame #4: 0x000000010600d87f watcher.node`std::__1::__map_iterator<std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event>, void*>*, long> >::operator++(this=0x00007ffeefbf6bd8, (null)=0) at map:808:9
    frame #5: 0x000000010600d18a watcher.node`EventList::toJS(this=0x00000001039354c0, env=0x00007ffeefbf6d80) at Event.hh:66:58
    frame #6: 0x0000000106010df7 watcher.node`Watcher::fireCallbacks(handle=0x0000000102e262c0) at Watcher.cc:104:44
    frame #7: 0x000000010093fb13 node`uv__async_io(loop=0x0000000101ce3080, w=<unavailable>, events=<unavailable>) at async.c:163:5 [opt]
    frame #8: 0x0000000100953fef node`uv__io_poll(loop=<unavailable>, timeout=-1) at kqueue.c:0 [opt]
    frame #9: 0x0000000100940141 node`uv_run(loop=0x0000000101ce3080, mode=UV_RUN_DEFAULT) at core.c:381:5 [opt]
    frame #10: 0x00000001000ddba5 node`node::NodeMainInstance::Run(this=0x00007ffeefbff490) at node_main_instance.cc:130:9 [opt]
    frame #11: 0x0000000100073295 node`node::Start(argc=<unavailable>, argv=<unavailable>) at node.cc:992:38 [opt]
    frame #12: 0x00007fff670a53d5 libdyld.dylib`start + 1
    frame #13: 0x00007fff670a53d5 libdyld.dylib`start + 1
```

[2]
```
Process 18582 stopped
* thread #8, stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
    frame #0: 0x0000000103a1b155 watcher.node`std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__is_long(this="") const at string:1421:22
   1418
   1419	    _LIBCPP_INLINE_VISIBILITY
   1420	    bool __is_long() const _NOEXCEPT
-> 1421	        {return bool(__r_.first().__s.__size_ & __short_mask);}
   1422
   1423	#if _LIBCPP_DEBUG_LEVEL >= 2
   1424
Target 0: (node) stopped.
Process 18582 launched: '/Users/mteegarden/dev/repos/node/node' (x86_64)
(llnode) bt
* thread #8, stop reason = EXC_BAD_ACCESS (code=EXC_I386_GPFLT)
  * frame #0: 0x0000000103a1b155 watcher.node`std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__is_long(this="") const at string:1421:22
    frame #1: 0x0000000103a10ab9 watcher.node`std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::size(this="") const at string:948:17
    frame #2: 0x0000000103a1af43 watcher.node`std::__1::enable_if<__can_be_converted_to_string_view<char, std::__1::char_traits<char>, std::__1::basic_string_view<char, std::__1::char_traits<char> > >::value, int>::type std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::compare<std::__1::basic_string_view<char, std::__1::char_traits<char> > >(this="", __t=0x0000700011267780) const at string:3681:23
    frame #3: 0x0000000103a36269 watcher.node`std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::compare(this="", __str="/Users/mteegarden/dev/repos/frontbucket/dist/parcel/present/panel-error.885c0998.js.948365684") const at string:3699:12
    frame #4: 0x0000000103a3621d watcher.node`bool std::__1::operator<<char, std::__1::char_traits<char>, std::__1::allocator<char> >(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)(__lhs="", __rhs="/Users/mteegarden/dev/repos/frontbucket/dist/parcel/present/panel-error.885c0998.js.948365684") at string:3924:18
    frame #5: 0x0000000103a361e1 watcher.node`std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >::operator(this=0x0000000102f06238, __x="", __y="/Users/mteegarden/dev/repos/frontbucket/dist/parcel/present/panel-error.885c0998.js.948365684")(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const at __functional_base:55:21
    frame #6: 0x0000000103a36185 watcher.node`std::__1::__map_value_compare<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event>, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, true>::operator(this=0x0000000102f06238, __x=0x4000000010b5ab0d, __y="/Users/mteegarden/dev/repos/frontbucket/dist/parcel/present/panel-error.885c0998.js.948365684")(std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) const at map:517:17
    frame #7: 0x0000000103a35fc0 watcher.node`std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event>, void*>*, long> std::__1::__tree<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event>, std::__1::__map_value_compare<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event>, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, true>, std::__1::allocator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event> > >::__lower_bound<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(this=0x0000000102f06228, __v="/Users/mteegarden/dev/repos/frontbucket/dist/parcel/present/panel-error.885c0998.js.948365684", __root=0x4000000010b5aaed, __result=0x000000010b8bba10) at __tree:2670:14
    frame #8: 0x0000000103a35e99 watcher.node`std::__1::__tree_iterator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event>, std::__1::__tree_node<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event>, void*>*, long> std::__1::__tree<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event>, std::__1::__map_value_compare<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event>, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, true>, std::__1::allocator<std::__1::__value_type<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event> > >::find<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >(this=0x0000000102f06228, __v="/Users/mteegarden/dev/repos/frontbucket/dist/parcel/present/panel-error.885c0998.js.948365684") at __tree:2599:20
    frame #9: 0x0000000103a35d0d watcher.node`std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, Event, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, Event> > >::find(this=0x0000000102f06228 size=0, __k="/Users/mteegarden/dev/repos/frontbucket/dist/parcel/present/panel-error.885c0998.js.948365684") at map:1375:68
    frame #10: 0x0000000103a34b73 watcher.node`EventList::update(this=0x0000000102f06228, path="/Users/mteegarden/dev/repos/frontbucket/dist/parcel/present/panel-error.885c0998.js.948365684") at Event.hh:35:26
    frame #11: 0x0000000103a34b01 watcher.node`EventList::create(this=0x0000000102f06228, path=<unavailable>) at Event.hh:30:20
    frame #12: 0x0000000103a6e579 watcher.node`FSEventsCallback(streamRef=0x0000000102e10fb0, clientCallBackInfo=0x0000000102f061e8, numEvents=17, eventPaths=0x000000010c0c59a0, eventFlags=0x0000000102dfe000, eventIds=0x0000000102dfd000) at FSEventsBackend.cc:65:13
    frame #13: 0x00007fff3c6d612e FSEvents`implementation_callback_rpc + 2991
    frame #14: 0x00007fff3c6d550a FSEvents`_Xcallback_rpc + 231
    frame #15: 0x00007fff3c6d5406 FSEvents`FSEventsD2F_server + 55
    frame #16: 0x00007fff3c6d7cbf FSEvents`FSEventsClientProcessMessageCallback + 43
    frame #17: 0x00007fff3b13032d CoreFoundation`__CFMachPortPerform + 246
    frame #18: 0x00007fff3b13022b CoreFoundation`__CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE1_PERFORM_FUNCTION__ + 41
    frame #19: 0x00007fff3b130189 CoreFoundation`__CFRunLoopDoSource1 + 527
    frame #20: 0x00007fff3b11817c CoreFoundation`__CFRunLoopRun + 2524
    frame #21: 0x00007fff3b11754e CoreFoundation`CFRunLoopRunSpecific + 455
    frame #22: 0x00007fff3b117362 CoreFoundation`CFRunLoopRun + 40
    frame #23: 0x0000000103a6f453 watcher.node`FSEventsBackend::start(this=0x0000000102f06608) at FSEventsBackend.cc:189:3
    frame #24: 0x0000000103a312a0 watcher.node`Backend::run(this=0x0000000102f064d8)::$_0::operator()() const at Backend.cc:79:7
    frame #25: 0x0000000103a3123d watcher.node`decltype(__f=0x0000000102f064d8)::$_0>(fp)()) std::__1::__invoke<Backend::run()::$_0>(Backend::run()::$_0&&) at type_traits:4361:1
    frame #26: 0x0000000103a311a5 watcher.node`void std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, Backend::run()::$_0>(__t=size=2, (null)=__tuple_indices<> @ 0x000070001126deb8)::$_0>&, std::__1::__tuple_indices<>) at thread:342:5
    frame #27: 0x0000000103a30a76 watcher.node`void* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, Backend::run()::$_0> >(__vp=0x0000000102f064d0) at thread:352:5
    frame #28: 0x00007fff672992eb libsystem_pthread.dylib`_pthread_body + 126
    frame #29: 0x00007fff6729c249 libsystem_pthread.dylib`_pthread_start + 66
    frame #30: 0x00007fff6729840d libsystem_pthread.dylib`thread_start + 13
```

### PR DESCRIPTION
**Pull request by DeMoorJasper**
This PR adds a locking behaviour to Events to make it a little more thread-safe, this should solve #39 

Ran the stress-test for 30 minutes and it didn't crash so I'm fairly positive this is fixed :)

Fixes #39

### CODE REVIEW

**Comment by DeMoorJasper** [comment]
It didn't fix it after all just took a bit longer before it crashed...

**Comment by mischnic** [comment]
Building with the address sanitizer might throw a more helpful error (`-g -fsanitize=address -fno-omit-frame-pointer` )

**Comment by DeMoorJasper** [comment]
@mischnic awesome, thanks for the tip. More helpful errors would definitely be helpful

File: src/Event.hh [hunk 1]
```diff
@@ -27,51 +27,57 @@ struct Event {
 class EventList {
 public:
   void create(std::string path) {
-    Event *event = update(path);
+    std::lock_guard<std::mutex> l(mMutex);
```
Comment (DeMoorJasper) [inline src/Event.hh:30]: The locking in this file might be a bit overkill, we probably don't really need it anymore as the real issue was in watcher reassigning and reading the events array/map at the same time

**Pull request merged**